minicluster/src/main/java/alluxio/master/MultiMasterEmbeddedJournalLocalAlluxioCluster.java

```diff
@@ -0,0 +1,310 @@
+        try {
+          LOG.info("master {} is a standby. stopping it...", k);
+          mMasters.get(k).stop();
+          LOG.info("master {} stopped.", k);
+        } catch (Exception e) {
+          LOG.error(e.getMessage(), e);
+          return false;
+        }
+        return true;
+      }
+    }
+    return false;
+  }
+
+  /**
+   * Iterates over the masters in the order of master creation, stops the leader master.
+   *
+   * @return true if the leader master is successfully stopped, false otherwise
+   */
+  public boolean stopLeader() {
+    int leaderId = getLeaderIndex();
+    try {
+      LOG.info("master {} is the leader. stopping it...", leaderId);
+      getLocalAlluxioMasterByIndex(leaderId).stop();
+      LOG.info("master {} stopped.", leaderId);
+    } catch (Exception e) {
+      LOG.error(e.getMessage(), e);
+      return false;
+    }
+    return true;
+  }
+
+  /**
+   * Waits for the primary master to start until a timeout occurs.
+   *
+   * @param timeoutMs the number of milliseconds to wait before giving up and throwing an exception
+   */
+  public void waitForPrimaryMasterServing(int timeoutMs)
+      throws TimeoutException, InterruptedException {
+    CommonUtils.waitFor("the primary leader master to start",
+        () -> {
+          int leaderId = getLeaderIndex();
+          if (leaderId == -1) {
+            return false;
+          }
+          try {
+            getLocalAlluxioMasterByIndex(leaderId).getClient().listStatus(
+                new AlluxioURI("/"));
+            return true;
+          } catch (Exception e) {
+            return false;
+          }
+        },
+        WaitForOptions.defaults().setTimeoutMs(timeoutMs));
+  }
+
+  @Override
+  protected void startMasters() throws IOException {
+    for (int k = 0; k < mNumOfMasters; k++) {
+      Configuration.set(PropertyKey.MASTER_METASTORE_DIR,
+          PathUtils.concatPath(mWorkDirectory, "metastore-" + k));
+      MasterNetAddress address = mMasterAddresses.get(k);
+      Configuration.set(PropertyKey.LOGGER_TYPE, "MASTER_LOGGER");
+      Configuration.set(PropertyKey.MASTER_HOSTNAME, address.getHostname());
+      Configuration.set(PropertyKey.MASTER_RPC_PORT, address.getRpcPort());
+      Configuration.set(PropertyKey.MASTER_WEB_PORT, address.getWebPort());
+      Configuration.set(PropertyKey.MASTER_EMBEDDED_JOURNAL_PORT,
+          address.getEmbeddedJournalPort());
+      // foobar
+      Configuration.set(PropertyKey.MASTER_JOURNAL_FOLDER, mJournalFolders.get(k));
+
+      final LocalAlluxioMaster master = LocalAlluxioMaster.create(mWorkDirectory, false);
+      master.start();
+      LOG.info("master NO.{} started, isServing: {}, address: {}", k, master.isServing(),
+          master.getAddress());
+      mMasters.add(master);
+      // Each master should generate a new port for binding
+      Configuration.set(PropertyKey.MASTER_RPC_PORT, 0);
+      Configuration.set(PropertyKey.MASTER_WEB_PORT, 0);
```

Comment:
why is this?

Reply:
0 means using a default endpoint. This is copied from MultiMasterJournalLocalAlluxioCluster.java

Reply:
though i removed these two as they will be designated later when the master starts